File: internal/app/repo/postgres/report_repo.go
package postgres

import (
	"context"
	"database/sql"
	"fmt"
	"log/slog"
	"mahir-trade-be/internal/app/models"
	"mahir-trade-be/internal/app/repo/postgres/queries"

	"go.uber.org/dig"
)

type (
	ReportRepo interface {
		CreateReport(ctx context.Context, req models.Report) (id int64, err error)
		GetReports(ctx context.Context, req models.PaginationRequest) (reports []models.Report, totalCount int64, err error)
		GetReportByID(ctx context.Context, id int64) (report models.Report, err error)
		UpdateReport(ctx context.Context, req models.Report) (err error)
		SoftDeleteReport(ctx context.Context, id int64, deletedBy string) (err error)
	}

	ReportRepoImpl struct {
		dig.In

		*sql.DB
	}
)

func NewReportRepo(impl ReportRepoImpl) ReportRepo {
	return &impl
}

func (r *ReportRepoImpl) CreateReport(ctx context.Context, req models.Report) (id int64, err error) {
	rows, err := r.QueryContext(ctx, queries.QueryCreateReport, req.ReportName, req.ReportThumbnailURL, req.Contents, req.CreatedBy)
	if err != nil {
		slog.ErrorContext(ctx, fmt.Sprintf("error while CreateReport err: %v", err.Error()))
		return id, err
	}

	defer rows.Close()

	for rows.Next() {
		err = rows.Scan(&id)
		if err != nil {
			slog.ErrorContext(ctx, fmt.Sprintf("error while CreateReport err: %v", err.Error()))
			return id, err
		}
	}

	return id, nil
}

func (r *ReportRepoImpl) GetReports(ctx context.Context, req models.PaginationRequest) (reports []models.Report, totalCount int64, err error) {
	if req.Limit == 0 {
		req.Limit = 10
	}

	offset := 0
	if req.Page > 1 {
		offset = int((req.Page - 1) * req.Limit)
	}

	query := queries.QueryGetReports
	queryParams := []interface{}{}

	if req.Search != "" {
		query += " AND report_name ILIKE '%' || $1 || '%'"
		queryParams = append(queryParams, req.Search)
	}

	query += fmt.Sprintf(" LIMIT $%d OFFSET $%d", len(queryParams)+1, len(queryParams)+2)
	queryParams = append(queryParams, req.Limit, offset)

	rows, err := r.QueryContext(ctx, query, queryParams...)
	if err != nil {
		slog.ErrorContext(ctx, fmt.Sprintf("error while GetReports err: %v", err.Error()))
		return reports, totalCount, err
	}

	defer rows.Close()

	for rows.Next() {
		var report models.Report
		err = rows.Scan(&totalCount, &report.ID, &report.ReportName, &report.ReportThumbnailURL, &report.Contents, &report.CreatedBy, &report.UpdatedBy, &report.CreatedAt, &report.UpdatedAt)
		if err != nil {
			slog.ErrorContext(ctx, fmt.Sprintf("error while GetReports err: %v", err.Error()))
			return reports, totalCount, err
		}

		reports = append(reports, report)
	}
	if err != nil {
		slog.ErrorContext(ctx, fmt.Sprintf("error while marshalling reports err: %v", err.Error()))
		return reports, totalCount, err
	}
	return reports, totalCount, nil
}

func (r *ReportRepoImpl) GetReportByID(ctx context.Context, id int64) (report models.Report, err error) {
	row := r.QueryRowContext(ctx, queries.QueryGetReportByID, id)

	err = row.Scan(&report.ID, &report.ReportName, &report.ReportThumbnailURL, &report.Contents, &report.CreatedBy, &report.UpdatedBy, &report.CreatedAt, &report.UpdatedAt)
	if err != nil {
		if err == sql.ErrNoRows {
			return report, fmt.Errorf("report not found")
		}

		slog.ErrorContext(ctx, fmt.Sprintf("error while GetReportByID err: %v", err.Error()))
		return report, err
	}

	return report, nil
}

func (r *ReportRepoImpl) UpdateReport(ctx context.Context, req models.Report) (err error) {
	_, err = r.ExecContext(ctx, queries.QueryUpdateReport, req.ReportName, req.ReportThumbnailURL, req.Contents, req.UpdatedBy, req.ID)
	if err != nil {
		slog.ErrorContext(ctx, fmt.Sprintf("error while UpdateReport err: %v", err.Error()))
		return err
	}

	return nil
}

func (r *ReportRepoImpl) SoftDeleteReport(ctx context.Context, id int64, deletedBy string) (err error) {
	_, err = r.ExecContext(ctx, queries.QuertSoftDeleteReport, deletedBy, deletedBy, id)
	if err != nil {
		slog.ErrorContext(ctx, fmt.Sprintf("error while SoftDeleteReport err: %v", err.Error()))
		return err
	}

	return nil
}

File: src/repo/report_repo.ts
import { Pool } from "pg";
import { Report } from "../models/report.model";

import db from "../db/db.config";
import { ReportQueries } from "./queries/report_queries";
import { PaginationRequest } from "@utils/response";

export class ReportRepository {
  private pg: Pool;

  constructor() {
    this.pg = db;
  }

  // --- CREATE REPORT ---
  async createReport(req: Report): Promise<number> {
    try {
      const result = await this.pg.query(ReportQueries.CreateReport, [
        req.report_name,
        req.report_thumbnail_url,
        req.report_file_url,
        req.created_by,
      ]);

      if (result.rowCount === 0) {
        throw new Error("failed to create report");
      }

      return result.rows[0].id as number;
    } catch (err: any) {
      console.error(`[repo][report][createReport] error: ${err.message}`);
      throw err;
    }
  }

  // --- GET REPORTS (with pagination + search) ---
  async getReports(
    req: PaginationRequest
  ): Promise<{ reports: Report[]; totalCount: number }> {
    try {
      const limit = req.limit || 10;
      const offset = req.page > 1 ? (req.page - 1) * limit : 0;

      let query = ReportQueries.GetReports;
      const params: any[] = [];

      if (req.search) {
        query += ` AND report_name ILIKE '%' || $1 || '%'`;
        params.push(req.search);
      }

      query += ` LIMIT $${params.length + 1} OFFSET $${params.length + 2}`;
      params.push(limit, offset);

      const result = await this.pg.query(query, params);

      const reports: Report[] = (result.rows as any[]).map((row) => ({
        id: row.id,
        report_name: row.report_name,
        report_thumbnail_url: row.report_thumbnail_url,
        report_file_url: row.report_file_url,
        created_by: row.created_by,
        updated_by: row.updated_by,
        created_at: row.created_at,
        updated_at: row.updated_at,
      }));

      const totalCount =
        result.rows.length > 0 ? parseInt(result.rows[0].total_count, 10) : 0;

      return { reports, totalCount };
    } catch (err: any) {
      console.error(`[repo][report][getReports] error: ${err.message}`);
      throw err;
    }
  }

  // --- GET REPORT BY ID ---
  async getReportByID(id: number): Promise<Report> {
    try {
      const result = await this.pg.query(ReportQueries.GetReportByID, [id]);

      if (result.rowCount === 0) {
        throw new Error(`report with id ${id} not found`);
      }

      const row = result.rows[0];
      return {
        id: row.id,
        report_name: row.report_name,
        report_thumbnail_url: row.report_thumbnail_url,
        report_file_url: row.report_file_url,
        created_by: row.created_by,
        updated_at: row.updated_by,
        created_at: row.created_at,
        updated_by: row.updated_at,
      };
    } catch (err: any) {
      console.error(`[repo][report][getReportByID] error: ${err.message}`);
      throw err;
    }
  }

  // --- UPDATE REPORT ---
  async updateReport(req: Report): Promise<boolean> {
    try {
      const result = await this.pg.query(ReportQueries.UpdateReport, [
        req.report_name,
        req.report_thumbnail_url,
        req.report_file_url,
        req.updated_by,
        req.id,
      ]);

      return (result.rowCount ?? 0) > 0;
    } catch (err: any) {
      console.error(`[repo][report][updateReport] error: ${err.message}`);
      throw err;
    }
  }

  // --- SOFT DELETE REPORT ---
  async softDeleteReport(id: number, deletedBy: string): Promise<boolean> {
    try {
      const result = await this.pg.query(ReportQueries.SoftDeleteReport, [
        deletedBy,
        deletedBy,
        id,
      ]);

      return (result.rowCount ?? 0) > 0;
    } catch (err: any) {
      console.error(`[repo][report][softDeleteReport] error: ${err.message}`);
      throw err;
    }
  }
}
